### PR TITLE
Calibrate CJC at next boot

### PR DESCRIPTION
**Pull request by discip**
After reading this Ralim/IronOS#1387 post, I had the idea to make CJC (Cold Junction Compensation) calibration as accurate as possible by executing it at boot.

This is an attempt to implement that to solve Ralim/IronOS#1387.

Tested this on both **`TS80P`** & **`Pinecil`**.

#### What has changed:

1. **`Calibrate temperature`** is now called **`Calibrate CJC at next boot`**.
   1. It is now performed, as the name implies, at the next boot.
   2. The calibration will only start if the ***tip is installed*** & the ***temp difference between tip & handle is < 10°C*** !
   3. This setting is a single shot only (You have to reenable it after every reboot.).
2. **`Debug Menu`** was partially rearranged, aligned some entries were renamed.
    1. There is one more entry named **`Tip O`**, which indicates the actual CJC offset.
    2. **`CHan`** now **`Han C`**, is displayed directly in **`°C`** to match **`Tip C`** & **`Max C`**.
3. All (at least I hope so 😅) the corresponding documentations have been updated.

### CODE REVIEW

File: source/Core/Threads/GUIThread.cpp
Comment (discip) [inline source/Core/Threads/GUIThread.cpp:1006]: @Ralim
Even after putting in what you suggested it starts without the tip being attached.

Most likely I did do it the wrong way! 🤷‍♂️

**Comment by discip** [comment]
@Ralim
> I would yank this block of logic you have here out into a function like performTipOffsetCalibration()
>
> And then just call that if the function is set.

Unfortunately, I am not knowledgeable enough about programming to know what this should look like in code.
Maybe you have a comparable example? 😅

File: source/Core/Src/settingsGUI.cpp
Comment (discip) [inline source/Core/Src/settingsGUI.cpp:608]: @Ralim
I don't know what I destroyed here, but this is not working as anticipated! 😅

**Comment by discip** [comment]
@Ralim
Have you had any chance to assess the situation here?

File: source/Core/Threads/GUIThread.cpp
Comment (Ralim) [reply source/Core/Threads/GUIThread.cpp:1006]: >I would yank this block of logic you have here out into a function like performTipOffsetCalibration()..


I just meant to take the blob of code have added here and make it into its own function.
(yank just being slang for moving it up the file).

So i would have roughly:

```C
void performCJC(){
 while (preStartChecks() == 0) {
    ulTaskNotifyTake(pdTRUE, TICKS_100MS);
  }
 // Insert the rest of your code section here
}
```

Then where you currently have you code I would just do a 
```C

if (setting Is Set) {
performCJC();
}
```

File: source/Core/Src/settingsGUI.cpp
Comment (Ralim) [reply source/Core/Src/settingsGUI.cpp:608]: The draw Checkbox should be right. I would compare against `displayAdvancedSolderingScreens` & the advanced soldering toggle to check against.

Comment (discip) [reply source/Core/Src/settingsGUI.cpp:608]: Yes, the drawCheckbox works flawlessly,  but my attempt to make the **`SettingsCalibrationWarning`** to be displayed before the checkbox gets set, does not work!

Actually I was referring to this part of code:
```C#
static bool setCalibrate(void) {

  if (userConfirmation(translatedString(Tr->SettingsCalibrationWarning))) {
    // User confirmed
    // So we now perform the actual calculation
    setTipOffset();
    // So we now set the tick
    setSettingValue(SettingsOptions::CalibrateCJC, 1);
  }
  return false;
}
```

File: source/Core/Threads/GUIThread.cpp
Comment (discip) [reply source/Core/Threads/GUIThread.cpp:1007]: I tried that but I was not able to get it working as I kept getting 2 errors no matter what I changed.

Comment (discip) [reply source/Core/Threads/GUIThread.cpp:1007]: I tried that but I was not able to get it to work as I kept getting 2 errors no matter what I changed.

Comment (discip) [reply source/Core/Threads/GUIThread.cpp]: Ok, that worked!
But still neither the ***missing tip condition*** nor the ***temperature condition*** are respected.

File: source/Core/Src/settingsGUI.cpp
Comment (discip) [reply source/Core/Src/settingsGUI.cpp:607]: I don't understand why this works:
```C#
static bool setResetSettings(void) {
  if (userConfirmation(translatedString(Tr->SettingsResetWarning))) {
    resetSettings();
    warnUser(translatedString(Tr->SettingsResetMessage), 10 * TICKS_SECOND);
  }
  return false;
}

static void displayResetSettings(void) {}
```

... but not this:
```C#
static bool setCalibrate(void) {
  if (userConfirmation(translatedString(Tr->SettingsCalibrationWarning))) {
    // User confirmed
    // So we now set the tick
    setSettingValue(SettingsOptions::CalibrateCJC, 1);
  }
  return false;
}

static void displayCalibrate(void) { OLED::drawCheckbox(getSettingValue(SettingsOptions::CalibrateCJC)); }
```

Comment (Ralim) [reply source/Core/Src/settingsGUI.cpp:607]: Honestly, I think that _should_ work.
Maybe double check git hash in debug menu to make sure it wasnt a bad flash?
What do you see?

File: source/Core/Threads/GUIThread.cpp
Comment (Ralim) [inline source/Core/Threads/GUIThread.cpp:103]:
```suggestion
  if (!isTipDisconnected() && Temp <= 32) {
```

File: source/Core/Src/settingsGUI.cpp
Comment (discip) [reply source/Core/Src/settingsGUI.cpp:607]: 1. I tried it several times and I am getting always the same result.
2. The Calibrate entry appears with an empty checkbox and after pressing A it just gets checked without the warning message.

Comment (Ralim) [reply source/Core/Src/settingsGUI.cpp:607]: Ah figured it out.
You need to change the `SettingsOptions::CalibrateCJC` to `SettingsOptions::SettingsOptionsLength` to make it not use the automatic settings toggle and instead call your setSettingValue() function

File: source/Core/Threads/GUIThread.cpp
Comment (discip) [reply source/Core/Threads/GUIThread.cpp]: I tried that before, but in combination with the calibration checkbox, which unfortunately didn't work either.

File: source/Core/Src/settingsGUI.cpp
Comment (Ralim) [reply source/Core/Src/settingsGUI.cpp:607]: You can still use `setSettingValue(SettingsOptions::CalibrateCJC, 0);` / `setSettingValue(SettingsOptions::CalibrateCJC, 1);` Normally.

Just to cut down on code side, if that value is set, it assumes it cant just tell the settings code to increment the value. Which cuts down on code duplication (and space) since most values just increment and wrap.

Comment (discip) [reply source/Core/Src/settingsGUI.cpp:607]: 1. Will try that after I have resolved the conflicts after the latest merges.
2. I don't quite get what exactly is meant by that, will have to read it a few more times.

Comment (Ralim) [reply source/Core/Src/settingsGUI.cpp:656]: So you still use `setSettingValue` to set the value to 0/1 for on/off as per normal.
But in the menu definition you need to swap out the value.
To get past this check https://github.com/Ralim/IronOS/blob/c99a1995202bb5b173925ad9eaa8e6a247f984aa/source/Core/Src/settingsGUI.cpp#L957

Comment (discip) [reply source/Core/Src/settingsGUI.cpp]: So this finally works as I intended it to do!

If the tick is not set, the **`warning message`** is displayed if **`A`** gets pressed.
After confirming with **`A`** the tick is set. If **`B`** is used, though, it will not  be set!

However, if the tick is already set, another press on **`A`** will remove the tick again.

File: source/Core/Threads/GUIThread.cpp
Comment (discip) [reply source/Core/Threads/GUIThread.cpp]:
```C#
if (!isTipDisconnected() && Temp <= 32) {
```
As expected this did not solve the 'condition-issue'.

Do you have any Idea what could be the cause?

Comment (discip) [reply source/Core/Threads/GUIThread.cpp]: @Ralim
```C#
if (!isTipDisconnected() && Temp <= 32) {
```
As expected this did not solve the 'condition-issue'.

Do you have any Idea what could be the cause?

Comment (discip) [reply source/Core/Threads/GUIThread.cpp]: Ok, that worked!
But still neither the ***missing tip condition*** nor the ***temperature condition*** are respected.

@Ralim
```C#
if (!isTipDisconnected() && Temp <= 32) {
```
As expected this did not solve the 'condition-issue'.

Do you have any Idea what could be the cause?

Maybe some **function** to *measure the temperature* is not loaded at this point in time?

**Comment by discip** [comment]
@Ralim
Finally it works!

Maybe you could give it a try with a **`TS100`** because as you know I don't own one.

**Comment by discip** [comment]
@Ralim
Finally it works!

Maybe you could give it a try with a **`TS100`** because as you know I don't own one.

btw: I don't mind if you squash the commits as you like. Since there are quite a bunch of them by now. 😊

**Comment by discip** [comment]
@Ralim
Finally it works!

Maybe you could give it a try with a **`TS100`** because as you know I don't own one.

However, I am not sure about the *'max temp'* the tip must not exceed in order to make the calibration happen, if that one is not to low (currently <33°C).

btw: I don't mind if you squash the commits as you like. Since there are quite a bunch of them by now. 😊

**Comment by discip** [comment]
@Ralim
Finally it works!

Maybe you could give it a try with a **`TS100`** because as you know I don't own one.

However, I am not sure about the *'max temp'* the tip must not exceed in order to make the calibration happen, if that one is not to low (currently <33°C).

btw: I don't mind if you squash the commits as you like. Since there are quite a bunch of them by now. 😊

**Comment by discip** [comment]
@Ralim
Finally it works!

Maybe you could give it a try with a **`TS100`** because as you know I don't own one.

However, I am not sure about the *'max temp'* the tip must not exceed in order to make the calibration happen, if that one is not to low (currently <33°C).

btw: I don't mind if you squash the commits as you like. Since there is quite a bunch of them by now. 😊

**Comment by Ralim** [comment]
>Maybe you could give it a try with a TS100 because as you know I don't own one.

I'll need to find mine 🤣 

>Maybe that's too low (currently <33°C).

Hmm; I think its an ok starting point, maybe what we could look at is (`tip temp` - `handle temp`) is less than 10C instead, since its that skew that we care about?


>btw: I don't mind if you squash the commits as you like. Since there is quite a bunch of them by now. blush

Yeah thats no drama, this button is great for this:
image

**Comment by discip** [comment]
> Hmm; I think its an ok starting point, maybe what we could look at is (`tip temp` - `handle temp`) is less than 10C instead, since its that skew that we care about?

Than let me try to implement this first. 😊

**Comment by discip** [comment]
> Hmm; I think its an ok starting point, maybe what we could look at is (`tip temp` - `handle temp`) is less than 10C instead, since its that skew that we care about?

Than let me try to implement this first. 😊

Do we already use a similar function somewhere?

**Comment by discip** [comment]
> Hmm; I think its an ok starting point, maybe what we could look at is (`tip temp` - `handle temp`) is less than 10C instead, since its that skew that we care about?

Than let me try to implement this first. 😊

Do we already use a similar *'function'* somewhere?

**Comment by Ralim** [comment]
It would be roughly

(getTipTempX10()/10)-getHandleTemp())<10

*on phone, not sure exact function names. But both get called in debug menu for a pattern to copy

**Comment by discip** [comment]
I am at it right now! 😊

**Comment by discip** [comment]
@Ralim
It works!

But if the best calibration is achieved with the 'skew' being as small as possible, then I suggest it be less than 2°C.

What is your opinion on that?

**Comment by discip** [comment]
@Ralim
While I spent some time in the **`Debug Menu`** adapting things, I was not able to figure out how to align all the values.

If I wanted to move the values horizontally, would I have to put:
```C#
OLED::setCursor(x, 8);
```
in front of every:
```C#
OLED::printNumber( ...
```
### ?

Or what system is this based on?

**Comment by discip** [comment]
@Ralim
While I spent some time in the **`Debug Menu`** adapting things, I was not able to figure out how to align all the values.

If I wanted to move the values horizontally, would I have to put:
```C#
OLED::setCursor(x, 8);
```
in front of every:
```C#
OLED::printNumber( ...
```
### ?

**Comment by Ralim** [comment]
The left position of the number is just straight after the string for the menu finishes. However most are not drawing the leading "0"s. I don't think there is a left aligned number print mode. Iirc

**Comment by discip** [comment]
I do not understand how they are aligned left but if the value increases they don't shift right.

**Comment by Ralim** [comment]
So the value is printed right aligned, but its position is determined left aligned.

So, for example with a fixed size of 5

12345 -> "12345"
123-> "  123"
1 -> "    1"

So you wont make them line up on the left edge, as they will "hug" the right end of the screen by design

**Comment by discip** [comment]
> So, for example with a fixed size of 5

Is the **`5`** in this example what you are referring to?:

```C#
    case 7:
      // Temp in C
      OLED::printNumber(TipThermoModel::getTipInC(), 5, FontStyle::SMALL);
      break;
```

> 12345 -> "12345"
> 123-> " 123"
> 1 -> " 1"

Sorry but I do not understand that. 🤪😅

**Comment by Ralim** [comment]
yeah thats the 5, since all numbers are printed in a "fixed width" area of the screen. But they are printed right aligned inside that area.

**Comment by discip** [comment]
And this area is than attached right behind the according string?

**Comment by discip** [comment]
And this area is attached right behind the according string?

**Comment by Ralim** [comment]
The left edge of the area is attached to the end of the string. The right edge is defined by the number we use in the printNumber call

**Comment by discip** [comment]
I am already on it. 😊
Will report if it worked as soon as I am done.

**Comment by Ralim** [comment]
Inside `printVoltage()` will handle the formatting as that includes the `V`.

**Comment by discip** [comment]
> Inside `printVoltage()` will handle the formatting as that includes the `V`.

Is there a way to shift it to the right somehow?

**Comment by discip** [comment]
@Ralim
Adding spaces:
https://github.com/discip/IronOS/commit/9c15795c670667b343a83e412ba44de1fdd512a7

Makes the **`TS80P`** build fail:
https://github.com/discip/IronOS/actions/runs/2982676519

Do you have any clue on why only the **`TS80P`** is affected?

**Comment by Ralim** [comment]
Out of space, those extra bytes pushed you over the space left in flash

**Comment by discip** [comment]
Space seems to be my biggest enemy here! 😭😫😩

**Comment by discip** [comment]
Do you have any idea what could be done to give us some space for the **`TS80P`**?

**Comment by discip** [comment]
I think I am done here (this PR).

This can be merged now. 😊

**Comment by Ralim** [comment]
>btw: Do you have any idea what could help us to reclaim some space again?

Most likely future features will be masked off from the Miniware Irons (TS100/TS80/TS80P) at this point
There are things I would like to do, but they are small changes, there isnt much that would free up a significant chunk.

I want to look into compressing all languages (rather than just multi-lang packs) but not sure if that net gains any space or not.

**Review by Ralim — APPROVED** [review]
This looks good to me.
I'll hold off merging until I get a sec to test

**Comment by discip** [comment]
I am good with that! 😃

**Comment by discip** [comment]
@Ralim
I had to update the **`Debug Menu`** once more to give the alignment a chance.
Text is aligned left and numbers so that the integers are on par.

Hope you don't mind.

**Comment by discip** [comment]
@Ralim
I had to update the **`Debug Menu`** once more to give the alignment a chance.
Text, except for **`ID`**, is aligned left and numbers so that the integers are on par.

Hope you don't mind.

**Comment by discip** [comment]
@Ralim
I had to update the **`Debug Menu`** once more to give the alignment a chance.
Text is aligned left and numbers so that the integers are on par.

Hope you don't mind.

**Comment by discip** [comment]
@Ralim
Just a reminder:
Please don't forget to check if this also works for the **`TS100`** & the **`Pinecil v2`** before merging.

Thank you

**Comment by discip** [comment]
A side note I would like you to consider:

I don't know if it's just me, but I've never felt the need to use the ***button locking mode***.

Maybe we could just remove it? 🤷‍♂️ I've already done this in one of the branches of my fork.
So, creating a PR is no hurdle. 😊

**Comment by discip** [comment]
1. To save space I thought about removing unused features and found that I had never even tried the **`locking mode`**!
Do you know if this is a crucial feature or if someone would miss it if it was removed?

   In one of my branches of my fork I did exactly that and I got ~**80 KB** back.

2. And by removing the **`invert screen`** option in addition to the previous one, we save another ~**30 KB**.

So creating a PR would not be an issue. 😊

3. > I want to look into compressing all languages (rather than just multi-lang packs) but not sure if that net gains any space or not.

   1. Is there anything I could help you with on that?
   2. Or do you think that involves to much in-depth coding? 
   3. Maybe you could point me into the right direction?

**Comment by Ralim** [comment]
I do not want to remove locking mode as it is not a large amount of space, not sure about the 80KB figure?
It is also a reasonably new contribution hmm

Liekwise invert screen is liked by some so prefer not to take out the option, but might need to figure out about reducing its size cost?

**Comment by Ralim** [comment]
For the language option, I would need to figure out how the current system really works (I didnt write it)
IT may be as simple as removing the non-compressed path in the makefile but reallllllllly not sure off the top of my head

**Comment by discip** [comment]
> I do not want to remove locking mode ...

I'm fine with that. Just wanted to throw in my 2 cents. 😅

> ... not sure about the 80KB figure

Sorry, my bad! I used the size of the ***whole zip file*** as reference! 🤦‍♂️
So, for eg: *TS80P_EN.dfu* it saves about **511 bytes** or **687 bytes** respectively.


> ... IT may be as simple as removing the non-compressed path in the makefile ...

Ok, then I guess I just have to wait until you have time, because that alone is beyond my understanding! 🤣

**Comment by discip** [comment]
1. > I do not want to remove locking mode ...

   I'm fine with that. Just wanted to throw in my 2 cents. 😅

   > ... not sure about the 80KB figure

   Sorry, my bad! I used the size of the ***whole zip file*** as reference! 🤦‍♂️
   So, for eg: *TS80P_EN.dfu* it saves about **511 bytes** or **687 bytes** respectively.


2. > ... IT may be as simple as removing the non-compressed path in the makefile ...

   Ok, then I guess I just have to wait until you have time, because that alone is beyond my understanding! 🤣

**Comment by discip** [comment]
1. > I do not want to remove locking mode ...

   I'm fine with that. Just wanted to throw in my 2 cents. 😅

   > ... not sure about the 80KB figure

   Sorry, my bad! I used the size of the ***whole zip file*** as reference! 🤦‍♂️
   So, for eg: *TS80P_EN.dfu* it saves about **511 bytes** or **687 bytes** respectively.


2. > ... IT may be as simple as removing the non-compressed path in the makefile ...

   Ok, then I guess I just have to wait until you have time, because that alone is beyond my understanding! 🤣
3. Have you had a chance to look into the changes so far?

**Comment by Ralim** [comment]
Tested on TS100 and works perfectly, so I think its good to gooooooo

**Comment by discip** [comment]
@Ralim
I guess the *'bug-tag'* can now be removed from https://github.com/Ralim/IronOS/issues/1387#issuecomment-1226583547.

**Pull request merged**
